src/coreclr/jit/emitarm64.cpp

```diff
@@ -7677,14 +7679,20 @@ void emitter::emitIns_R_S(instruction ins, emitAttr attr, regNumber reg1, int va
         }
     }
 
-    // Is the ldr/str even necessary?
-    if (emitComp->opts.OptimizationEnabled() && IsRedundantLdStr(ins, reg1, reg2, imm, size, fmt))
+    assert(fmt != IF_NONE);
+
+    // Try to optimize a load/store with an alternative instruction.
+    if (isLdrStr && emitComp->opts.OptimizationEnabled() && OptimizeLdrStr(ins, attr, reg1, reg2, imm, size, fmt))
     {
+        // Ensure local variables are tracked.
+        emitLastIns->idAddr()->iiaLclVar.initLclVarAddr(varx, offs);
+        emitLastIns->idSetIsLclVar();
+#ifdef DEBUG
+        emitLastIns->idDebugOnlyInfo()->idVarRefOffs = emitVarRefOffs;
+#endif
```

Comment:
I don't think this code should live here; it should live in the optimization itself. This code shouldn't be making any assumptions about what the optimization did (and what it didn't do and needs additional fixing up), and the optimization should be more self-contained. Can you pass the varx/offs to the optimization function?

Reply:
> I would propose as a first step, to try and get this PR merged sooner rather than later, to simply not do this optimization if either the previous or current instruction contains lclvar info. Yes, this might remove a lot of the wins. Adding appropriate support for the single or double lclvar info cases could be added back as a follow-up.

Done.  Looks like this has removed about 60% of the gains. Which is a shame, but, I agree, let's make this safe and get it merged sooner rather than later.

I'm still allowing the optimisation happen for `IsRedundantLdStr()`, given that's how it works in `HEAD` today, so assuming that is safe.

> To support this, I think we need a generalized instrDesc that can store 2 separate lclvar infos. 

That sounds reasonable, but definitely for a follow on PR.

> fwiw, I brought up this problem back in November: [#77540 (comment)](https://github.com/dotnet/runtime/pull/77540#issuecomment-1299600281)

Was looking for GC comments yesterday but missed that one. :/